cost_migration.go

```diff
@@ -0,0 +1,168 @@
+	}
+
+	var (
+		cost                 loopdb.SwapCost
+		swapPaid, prepayPaid bool
+	)
+
+	// Now that we have the prepay and swap amount, we can calculate the
+	// total cost of the swap. Note that we only need to account for the
+	// server cost in case the swap was successful or if the sweep timed
+	// out. Otherwise the server didn't pull the off-chain htlc not the
+	// prepay.
+	switch loopOutSwap.State().State {
+	case loopdb.StateSuccess:
+		cost.Server = swapPaymentAmount + prepayAmount -
+			loopOutSwap.Contract.AmountRequested
+
+		swapPaid = true
+		prepayPaid = true
+
+	case loopdb.StateFailSweepTimeout:
+		cost.Server = prepayAmount
+
+		prepayPaid = true
+
+	default:
+		cost.Server = 0
+	}
+
+	// Now attempt to look up the actual payments so we can calculate the
+	// total routing costs.
+	prepayPaymentFee, ok := paymentFees[hash]
+	if prepayPaid && ok {
+		cost.Offchain += prepayPaymentFee.ToSatoshis()
+	} else {
+		log.Debugf("Prepay payment %s is missing, won't account for "+
+			"routing fees", hash)
+	}
+
+	swapPaymentFee, ok := paymentFees[swapHash]
+	if swapPaid && ok {
+		cost.Offchain += swapPaymentFee.ToSatoshis()
+	} else {
+		log.Debugf("Swap payment %s is missing, won't account for "+
+			"routing fees", swapHash)
+	}
+
+	// For the on-chain cost, just make sure that the cost is positive.
+	cost.Onchain = loopOutSwap.State().Cost.Onchain
+	if cost.Onchain < 0 {
+		cost.Onchain *= -1
+	}
+
+	return cost, nil
+}
+
+// MigrateLoopOutCosts will calculate the correct cost for all loop out swaps
+// and override the cost values of the last update in the database.
+func MigrateLoopOutCosts(ctx context.Context, lnd lndclient.LndServices,
+	db loopdb.SwapStore) error {
+
+	migrationDone, err := db.HasMigration(ctx, costMigrationID)
+	if err != nil {
+		return err
+	}
+	if migrationDone {
+		log.Infof("Cost cleanup migration already done, skipping")
+
+		return nil
+	}
+
+	log.Infof("Starting cost cleanup migration")
+	startTs := time.Now()
+	defer log.Infof("Finished cost cleanup migration in %v",
+		time.Since(startTs))
+
+	// First we'll fetch all loop out swaps from the database.
+	loopOutSwaps, err := db.FetchLoopOutSwaps(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Next we fetch all payments from LND.
+	payments, err := lnd.Client.ListPayments(
+		ctx, lndclient.ListPaymentsRequest{},
+	)
+	if err != nil {
+		return err
+	}
+
+	// Gather payment fees to a map for easier lookup.
+	paymentFees := make(map[lntypes.Hash]lnwire.MilliSatoshi)
+	for _, payment := range payments.Payments {
+		paymentFees[payment.Hash] = payment.Fee
+	}
+
+	// Now we'll calculate the cost for each swap and finally update the
+	// costs in the database.
+	updatedCosts := make(map[lntypes.Hash]loopdb.SwapCost)
+	for _, loopOutSwap := range loopOutSwaps {
+		cost, err := CalculateLoopOutCost(
+			lnd.ChainParams, loopOutSwap, paymentFees,
+		)
+		if err != nil {
+			return err
+		}
+		updatedCosts[loopOutSwap.Hash] = cost
```

Comment:
May check that `updatedCosts` doesn't have the key yet.

Reply:
Added.